### PR TITLE
Extended schema and added metrics monitor.

### PR DESCRIPTION
**Pull request by carlosgarciaalvarado**
Summary:

* This change extends the JSON schema to allow users to define index fields.
* Also a metrics class is added to allow users to provide a custom monitor.

### CODE REVIEW

File: src/main/java/io/confluent/connect/elasticsearch/Mapping.java [hunk 1]
```diff
@@ -100,11 +97,28 @@ public static JsonObject getMapping(JestClient client, String index, String type
    * Infer mapping from the provided schema.
    * @param schema The schema used to infer mapping.
    */
-  public static JsonNode inferMapping(Schema schema, Map<String, String> fieldMapping) {
+  public static JsonNode inferMapping(Schema schema, IndexConfigurationProvider indexConfigurationProvider) {
+    return inferMapping(schema, indexConfigurationProvider, 0);
+  }
+
+  /**
+   * Infer mapping from the provided schema with nesting level.
```
Comment (joykent99) [inline src/main/java/io/confluent/connect/elasticsearch/Mapping.java]: Can you elaborate what nesting level is used here?

File: src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java [hunk 2]
```diff
@@ -335,9 +350,12 @@ public BulkResponse call() throws Exception {
 
     private BulkResponse execute() throws Exception {
       final B bulkReq;
+      Instant startInstant = clock.instant();
+
       try {
         bulkReq = bulkClient.bulkRequest(batch);
       } catch (Exception e) {
+        insertLatency(startInstant);
```
Comment (joykent99) [inline src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java]: Should this be in the finally block?

Comment (joykent99) [reply src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java]: Or should we have two latencies, one for successful insert, and the other for failed ones?

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java [hunk 3]
```diff
@@ -140,7 +141,10 @@ protected static ConfigDef baseConfigDef() {
                   + "Valid values are `" + CREATE_INDEX_AT_OPEN + "` or `" + CREATE_INDEX_AT_WRITE + "`.")
           .define(CUSTOM_DOCUMENT_TRANSFORMER_CONFIG, Type.CLASS, null, Importance.LOW,
                   "The class to use to perform any final transformations on documents after conversion to JSON.",
-                  group, ++order, Width.SHORT, "Custom Document Transformer Class");
+                  group, ++order, Width.SHORT, "Custom Document Transformer Class")
+          .define(CUSTOM_METRICS_CONFIG, Type.CLASS, null, Importance.LOW,
+              "The class to use to gather Elasticsearch metrics.",
+              group, ++order, Width.LONG, "Custom Metrics Class");
```
Comment (elalonde) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkConnectorConfig.java]: I love this approach.

File: src/main/java/io/confluent/connect/elasticsearch/Mapping.java [hunk 4]
```diff
@@ -142,27 +157,71 @@ public static JsonNode inferMapping(Schema schema, Map<String, String> fieldMapp
...
           if (fieldMapping != null && fieldMapping.containsKey(fieldName)) {
+
+            String value = fieldMapping.get(fieldName);
+
             ObjectNode customTypeNode = JsonNodeFactory.instance.objectNode();
-            customTypeNode.set("type",
-                  JsonNodeFactory.instance.textNode(fieldMapping.get(fieldName)));
+            String customFieldName = fieldMapping.get(fieldName);
+
+            if (value.contains("@")) {
```
Comment (elalonde) [inline src/main/java/io/confluent/connect/elasticsearch/Mapping.java]: A comment here explaining why @ is special (the reader may not know)

Comment (carlosgarciaalvarado) [reply src/main/java/io/confluent/connect/elasticsearch/Mapping.java]: Done.

File: src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java (hunk 2 above)
Comment (carlosgarciaalvarado) [reply src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java]: Today we only have one, I tried not to change the metrics behavior in this patch (I can modify it in a future patch if this gets too misleading.)

In regards to the Finally, I cannot add it to a "Finally block" given that I only want this metric to be captured if the request failed or after a certain amount of retries have been executed (in the second try/catch block.)

**Pull request merged**